pkg/pipeline/hooks/drivers/bitbucket.go

```diff
@@ -0,0 +1,143 @@
+package drivers
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/rancher/rancher/pkg/pipeline/providers"
+	"github.com/rancher/rancher/pkg/pipeline/remote/bitbucketcloud"
+	"github.com/rancher/rancher/pkg/pipeline/remote/model"
+	"github.com/rancher/rancher/pkg/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	BitbucketCloudWebhookHeader  = "X-Hook-UUID"
+	bitbucketCloudEventHeader    = "X-Event-Key"
+	bitbucketCloudPushEvent      = "repo:push"
+	bitbucketCloudPrCreatedEvent = "pullrequest:created"
+	bitbucketCloudPrUpdatedEvent = "pullrequest:updated"
+	bitbucketCloudStateOpen      = "OPEN"
+)
+
+type BitbucketCloudDriver struct {
+	PipelineLister             v3.PipelineLister
+	PipelineExecutions         v3.PipelineExecutionInterface
+	SourceCodeCredentials      v3.SourceCodeCredentialInterface
+	SourceCodeCredentialLister v3.SourceCodeCredentialLister
+}
+
+func (b BitbucketCloudDriver) Execute(req *http.Request) (int, error) {
+	event := req.Header.Get(bitbucketCloudEventHeader)
+	if event != bitbucketCloudPushEvent && event != bitbucketCloudPrCreatedEvent && event != bitbucketCloudPrUpdatedEvent {
+		return http.StatusUnprocessableEntity, fmt.Errorf("not trigger for event:%s", event)
+	}
+
+	pipelineID := req.URL.Query().Get("pipelineId")
+	ns, name := ref.Parse(pipelineID)
+	pipeline, err := b.PipelineLister.Get(ns, name)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return http.StatusUnprocessableEntity, err
+	}
+	logrus.Infof("get body:%s", string(body))
+
+	info := &model.BuildInfo{}
+	if event == bitbucketCloudPushEvent {
+		info, err = parseBitbucketPushPayload(body)
+		if err != nil {
+			return http.StatusUnprocessableEntity, err
+		}
+	} else if event == bitbucketCloudPrCreatedEvent || event == bitbucketCloudPrUpdatedEvent {
+		info, err = parseBitbucketPullRequestPayload(body)
+		if err != nil {
+			return http.StatusUnprocessableEntity, err
+		}
+	}
+
+	if (info.Event == utils.WebhookEventPush && !pipeline.Spec.TriggerWebhookPush) ||
+		(info.Event == utils.WebhookEventTag && !pipeline.Spec.TriggerWebhookTag) ||
+		(info.Event == utils.WebhookEventPullRequest && !pipeline.Spec.TriggerWebhookPr) {
+		return http.StatusUnavailableForLegalReasons, fmt.Errorf("trigger for event '%s' is disabled", info.Event)
+	}
+
+	pipelineConfig, err := providers.GetPipelineConfigByBranch(b.SourceCodeCredentials, b.SourceCodeCredentialLister, pipeline, info.Branch)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	if pipelineConfig == nil {
```

Comment:
it feels like a lot of code in bitbucket.go and bitbucket_server gets repeated; lets see if we can extract and re-use the common parts